[studio-ui] Preview Authorisation

aachavarria opened the pull request:
GlobalState.ts: adding permissions;
PreviewConcierge.tsx, ToolBar.tsx: adding permissions checks;
security.ts: creating getUserPermissions();
creating permissions epic,reducer, actions;
adding editModeChoice to localStorage;
ContextMenu.tsx: adding anchorOrigin;
common-api.js: updating DeleteContent createLegacyCallbackListener to support unsubscribe on cancel;
ContentLocalizationDialog.tsx: WIP pending work to move this to GlobalDialogManager.tsx;
CopyItemsDialog.tsx: Moving dialog to GlobalDialogManager.tsx; updating Dialog UI;
CreateFolderDialog.tsx: renaming, moving it to GlobalDialogManager.tsx;
PathSelectionDialog.tsx: Pending dispatch of showCreateFolderDialog;
ItemMenu.tsx: Refactor of ComponentMenu.tsx;
Updating pathNavigator.ts to use DetailedItem;
Widget.tsx: Updating widget to use ItemMenu.tsx;
Updating Dialogs to use DetailedItem;
Creating Permissions/DetailedItem/ClipBoard actions/reducer/epics;

### Ticket reference or full description of what's in the PR

aachavarria committed:
ContextMenu.tsx: adding anchorOrigin;
GlobalState.ts: adding permissions;
PreviewConcierge.tsx, ToolBar.tsx: adding permissions checks;
security.ts: creating getUserPermissions();
creating permissions epic,reducer, actions;
adding editModeChoice to localStorage;

aachavarria committed:
Merge branches 'feature/2296' and 'feature/permissions' of https://github.com/aachavarria/studio-ui into feature/permissions

aachavarria committed:
adding permissions to ComponentMenu.tsx

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/permissions

aachavarria committed:
common-api.js: updating DeleteContent createLegacyCallbackListener to support unsubscribe on cancel;
ContentLocalizationDialog.tsx: WIP pending work to move this to GlobalDialogManager.tsx;
CopyItemsDialog.tsx: Moving dialog to GlobalDialogManager.tsx; updating Dialog UI;
CreateFolderDialog.tsx: renaming, moving it to GlobalDialogManager.tsx;
PathSelectionDialog.tsx: Pending dispatch of showCreateFolderDialog;
ItemMenu.tsx: Refactor of ComponentMenu.tsx;
Updating pathNavigator.ts to use DetailedItem;
Widget.tsx: Updating widget to use ItemMenu.tsx;
Updating Dialogs to use DetailedItem;
Creating Permissions/DetailedItem/ClipBoard actions/reducer/epics;

aachavarria committed:
styles to the loading of ItemMenu.tsx;

aachavarria committed:
wcm-assets-folder.js: changes on the onClose Callback;
Reduxify BulkUploadDialog.tsx, CreateFileDialog.tsx;
Updating ItemMenu.tsx with createTemplate,createController, upload, ...etc options;

aachavarria committed:
adding template-editor.js readonly;

aachavarria committed:
Changes on Copy Service to receive path or copyItem;

aachavarria committed:
adding duplicateAsset/Item actions, epics;

aachavarria committed:
adding cancelUnsubscribe to createLegacyCallbackListener DISPATCH_DOM_EVENT

aachavarria committed:
adding to common-api.js: cancelUnsubscribe to createLegacyCallbackListener DISPATCH_DOM_EVENT;

aachavarria committed:
creating updateDetailedItem() action;

aachavarria committed:
updating fetchUserPermissions and fetchDetailedItem to remove site parameter;

aachavarria committed:
Creating PreviewDialog.tsx;
Adding ContentLoader to ItemMenu.tsx;
Adding icons and preview to pathNavigator;

aachavarria committed:
Adding PreviewDialog to Search.tsx

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -731,20 +732,29 @@ var nodeOpen = false,
 
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(eventIdCancel, () => {
+          unsubscribe();
+        });
```
For better runtime safety, let's declare the vars before assigning and using.
```js
let unsubscribe, cancelUnsubscribe;
unsubscribe = CrafterCMSNext...;
cancelUnsubscribe = CrafterCMSNext...;
```

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -840,11 +845,21 @@ var nodeOpen = false,
+
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(eventIdCancel, () => {
+          unsubscribe();
         });
```
Same as above

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -880,11 +896,21 @@ var nodeOpen = false,
+        });
+
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(eventIdCancel, () => {
+          unsubscribe();
```
Same as above

rart commented on static-assets/components/cstudio-common/common-api.js:
```diff
@@ -2559,11 +2586,21 @@ var nodeOpen = false,
+
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(eventIdCancel, () => {
+          unsubscribe();
         });
```
Same as above

rart commented on static-assets/components/cstudio-contextual-nav/quick-create.js:
```diff
@@ -124,14 +141,23 @@ CStudioAuthoring.ContextualNav.WcmQuickCreate = CStudioAuthoring.ContextualNav.W
+
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(contentTypeSelected, () => {
+          unsubscribe();
         });
```
As above

rart commented on static-assets/components/cstudio-contextual-nav/quick-create.js:
```diff
@@ -108,12 +119,18 @@ CStudioAuthoring.ContextualNav.WcmQuickCreate = CStudioAuthoring.ContextualNav.W
+
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(editDialogSuccess, (response) => {
+          unsubscribe();
+        });
```
As above

rart commented on static-assets/components/cstudio-contextual-nav/wcm-site-dropdown-mods/wcm-assets-folder.js:
```diff
@@ -2033,7 +2033,7 @@ var storage = CStudioAuthoring.Storage;
         bulkUpload.setAttribute('id', 'bulkUpload');
         document.documentElement.append(bulkUpload);
-        const onDismiss = (dropZoneStatus) => {
+        const onClose = ({ dropZoneStatus }) => {
```
What caused this change?

aachavarria replied:
because i did a change to return a response, onClose(response: { dropZoneStatus: DropZoneStatus, path?: string }):

rart commented on static-assets/components/cstudio-contextual-nav/wcm-site-dropdown-mods/wcm-root-folder.js:
```diff
@@ -3304,26 +3305,41 @@
         });
 
-        CrafterCMSNext.createLegacyCallbackListener(contentDialogTypeSelectedId, (response) => {
+        const unsubscribe = CrafterCMSNext.createLegacyCallbackListener(contentDialogTypeSelectedId, (response) => {
```
Same as above

rart commented on static-assets/components/cstudio-contextual-nav/wcm-site-dropdown-mods/wcm-root-folder.js:
```diff
@@ -3349,11 +3365,15 @@
           }
         });
 
+        const cancelUnsubscribe = CrafterCMSNext.createLegacyCallbackListener(contentTypeSelected, () => {
```
Same as above

rart commented on static-assets/components/cstudio-contextual-nav/wcm-site-dropdown-mods/wcm-root-folder.js:
```diff
@@ -3304,26 +3305,41 @@
             });
 
-            CrafterCMSNext.createLegacyCallbackListener(eventIdSuccess, (response) => {
+            const unsubscribe = CrafterCMSNext.createLegacyCallbackListener(eventIdSuccess, (response) => {
```
- Rename - above colliding names with the upper scope.
- Isn't this missing it's on "cancelUnsubscribe"?

rart commented on static-assets/components/cstudio-forms/template-editor.js:
```diff
@@ -782,14 +785,14 @@ CStudioAuthoring.Module.requireModule(
                             });
 
-                            CrafterCMSNext.createLegacyCallbackListener(createTemplateOnOk, () => {
+                            const unsubscribe = CrafterCMSNext.createLegacyCallbackListener(createTemplateOnOk, () => {
```
Same as above

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -62,7 +62,7 @@ export function parseQueryString() {
 
 // TODO: an initial path with trailing `/` breaks
-export function itemsFromPath(path: string, root: string, items: LookupTable<SandboxItem>): SandboxItem[] {
+export function itemsFromPath(path: string, root: string, items: LookupTable<DetailedItem>): DetailedItem[] {
```
In reality, all this needs is an item that has a `.path` prop, no? Try to loosen/adjust the type so we can use the method for different item types that match the deduction requirements.

rart commented on ui/app/src/utils/content.ts:
```diff
@@ -115,7 +118,7 @@ export function parseLegacyItemToBaseItem(item: LegacyItem): BaseItem {
     previewUrl: item.uri?.includes('index.xml') ? (item.browserUri || '/') : null,
     systemType: getLegacyItemSystemType(item),
-    mimeType: null,
+    mimeType: item.mimeType,
```
Please check the other `parseLegactyTo{*}Item` to include this too if applicable.

aachavarria replied:
the other ones already uses this method so it is included

rart commented on ui/app/src/utils/array.ts:
```diff
@@ -39,6 +41,12 @@ export function asArray<T = unknown>(value: T): T[] {
 
+export function createPresenceTable(list: string[]): LookupTable<boolean> {
+  const table = {};
+  list.forEach(value => table[value.replaceAll(' ', '_')] = true);
```
As a generic function, the `value.replace(...)` shouldn't be done here — looks too specific to your case or simply unnecessary.
What was that for? Ping me to discuss if you need help to adjust.

aachavarria replied:
fixed

rart commented on ui/app/src/state/epics/misc.ts:
```diff
@@ -0,0 +1,113 @@
+        })
+      );
+    })
+  );
```
These can go with content epics

rart commented on ui/app/src/state/actions/misc.ts:
```diff
@@ -20,3 +20,10 @@ import StandardAction from '../../models/StandardAction';
+export const changeContentType = createAction<{ contentTypeId: string, path: string }>('CHANGE_CONTENT_TYPE');
+export const editTemplate = createAction<{ contentTypeId: string }>('EDIT_TEMPLATE');
+export const itemDuplicate = createAction<{ path: string, onSuccess: StandardAction }>('ITEM_DUPlICATE');
+export const assetDuplicate = createAction<{ path: string, onSuccess: StandardAction }>('ASSET_DUPlICATE');
```
content actions

rart commented on ui/app/src/state/actions/content.ts:
```diff
@@ -16,9 +16,32 @@
+// region clipboard
+export const setClipBoard = createAction<{ path: string }>('SET_CLIPBOARD');
+export const unSetClipBoard = createAction('UNSET_CLIPBOARD');
+
```
// endregion

rart commented on ui/app/src/services/security.ts:
```diff
@@ -26,6 +27,14 @@ export function encrypt(text: string, site: string = ''): Observable<string> {
   );
 }
 
+export function getUserPermissions(site: string, path: string, user: string): Observable<any> {
```
Please type

rart commented on ui/app/src/services/content.ts:
```diff
@@ -1482,25 +1483,31 @@ export function cut(site: string, item: SandboxItem): Observable<any> {
 }
 
-export function getPages(site: string, item: any): Observable<any> {
+export function getPages(site: string, item: DetailedItem): Observable<any> {
```
- Return type
- Receive the path instead of the full item
- What's this for?

aachavarria replied:
this is used for CopyDialog, its returns the copy item and his childrens

rart commented on ui/app/src/modules/Preview/ToolBar.tsx:
```diff
@@ -285,6 +291,12 @@ export default function ToolBar() {
+  const permissions = useSelection((state) => state.content.permissions);
+  const write = permissions?.[item?.path]?.['write'];
+  const createContent = permissions?.[item?.path]?.['create_content'];
+  //endregion
```
Add space after comment opening

rart commented on ui/app/src/modules/Content/Dependencies/DependenciesDialog.tsx:
```diff
@@ -15,7 +15,7 @@
 
 import React, { PropsWithChildren, useCallback, useEffect, useMemo, useState } from 'react';
-import { SandboxItem } from '../../../models/Item';
+import { DetailedItem, SandboxItem } from '../../../models/Item';
```
- Please check if dialog can work with a BaseItem so that it could receive Sandbox or Detailed item alike.
- Looks like you've completely removed SandboxItem so make sure to clean imports

aachavarria replied:
we agreed to use detailed Item now.

aachavarria replied:
The dialogs should use the Detailed item state from redux, there are pending work for this.

rart commented on ui/app/src/modules/Content/Authoring/SingleItemSelector.tsx:
```diff
@@ -20,7 +20,7 @@ import IconButton from '@material-ui/core/IconButton';
 import clsx from 'clsx';
 import { Variant } from '@material-ui/core/styles/createTypography';
-import { SandboxItem } from '../../../models/Item';
+import { DetailedItem } from '../../../models/Item';
```
Please check if dialog can work with a BaseItem so that it could receive Sandbox or Detailed item alike.

rart commented on ui/app/src/models/Version.ts:
```diff
@@ -14,7 +14,7 @@
  */
 
-import { LegacyItem, SandboxItem } from './Item';
+import { DetailedItem, LegacyItem } from './Item';
```
Please check if this can work with a BaseItem so that it could receive Sandbox or Detailed item alike.

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -86,6 +93,11 @@ export interface GlobalState {
+    items: {
+      byId: LookupTable<DetailedItem>
+    };
+    clipboard?: string;
```
State props shouldn't be optional

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -86,6 +93,11 @@ export interface GlobalState {
       items: QuickCreateItem[];
     };
+    permissions: LookupTable<LookupTable<boolean>>;
+    items: {
```
There's a comment a couple of lines above also referencing items; remove.

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -86,6 +93,11 @@ export interface GlobalState {
     };
+    permissions: LookupTable<LookupTable<boolean>>;
+    items: {
+      byId: LookupTable<DetailedItem>
```
byPath

rart commented on ui/app/src/models/GetChildrenResponse.ts:
```diff
@@ -14,9 +14,9 @@
 
-export interface GetChildrenResponse extends Array<SandboxItem> {
-  parent: SandboxItem;
+export interface GetChildrenResponse extends Array<DetailedItem> {
```
The getChildrenBy* apis return sandbox items - i.e. not detailed.

rart commented on ui/app/src/components/Navigation/PathNavigator/utils.ts:
```diff
@@ -14,12 +14,33 @@
+  return item.systemType === 'folder';
+}
+
+export function getLoaderItems(item: DetailedItem): number {
```
- getNumOfMenuOptionsForItem
- Loosen type so it can be used for the different item types

rart commented on ui/app/src/components/Navigation/PathNavigator/utils.ts:
```diff
@@ -14,12 +14,33 @@
+    || item.systemType === 'taxonomy';
+}
+
+export function isFolder(item: DetailedItem): boolean {
```
Loosen type so it can be used for the different item types

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigatorBreadcrumbs.tsx:
```diff
@@ -14,7 +14,7 @@
  */
 
-import { SandboxItem } from '../../../models/Item';
+import { DetailedItem } from '../../../models/Item';
```
Can this work with BaseItem instead?

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigatorList.tsx:
```diff
@@ -15,27 +15,29 @@
 
 import List from '@material-ui/core/List';
-import { SandboxItem } from '../../../models/Item';
+import { DetailedItem } from '../../../models/Item';
```
Can this work with BaseItem instead?

rart commented on ui/app/src/components/Navigation/PathNavigator/Widget.tsx:
```diff
@@ -17,36 +17,20 @@
-} from '../../../services/content';
-import { getTargetLocales } from '../../../services/translation';
-import { LegacyItem, SandboxItem } from '../../../models/Item';
+import { DetailedItem } from '../../../models/Item';
```
Can this work with BaseItem instead?

rart reviewed (changes requested):
- [ ] Review & fix codacy code review issues
- [x] Let's discuss the use of DetailedItem vs SandboxItem

rart commented on ui/app/src/components/ContextMenu.tsx:
```diff
@@ -73,38 +69,62 @@ export default function ContextMenu(props: CustomMenuProps) {
       onClose={onClose}
+      anchorOrigin={anchorOrigin}
     >
+      <div>
+        <ContextMenuItems
+          sections={sections}
+          classes={classes}
+          onMenuItemClicked={onMenuItemClicked}
+        />
+      </div>
```
Is the wrapper div required?

aachavarria replied:
yes, because the issue with the foward ref

rart commented on ui/app/src/components/Dialogs/BulkUploadDialog.tsx:
```diff
@@ -628,18 +631,71 @@ export default function (props: BulkUploadProps) {
+      })}
+      onClose={() => onClose({ dropZoneStatus, path })}
+      fullWidth
+      maxWidth={'sm'}
```
Format

rart commented on ui/app/src/components/Dialogs/ContentLocalizationDialog.tsx:
```diff
@@ -34,7 +34,7 @@ import { useActiveSiteId, useUnmount } from '../../utils/hooks';
 import DialogHeader from './DialogHeader';
 import SingleItemSelector from '../../modules/Content/Authoring/SingleItemSelector';
-import { SandboxItem } from '../../models/Item';
+import { DetailedItem } from '../../models/Item';
```
Could this work with a BaseItem?

rart commented on ui/app/src/components/Dialogs/PathSelectionDialog.tsx:
```diff
@@ -261,7 +261,8 @@ function PathSelectionDialogWrapper(props: PathSelectionDialogProps) {
         </Button>
       </DialogFooter>
-      <CreateNewFolderDialog
+      {/* TODO: use Dispatch*/}
```
ToDo

aachavarria replied:
cant use dispatch because this is used on siteConfig

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigatorItem.tsx:
```diff
@@ -14,31 +14,33 @@
  */
 
-import { SandboxItem } from '../../../models/Item';
+import { DetailedItem } from '../../../models/Item';
```
Review usage of BaseItem instead

aachavarria committed:
declare unsubscribe, cancelUnsubscribe before assign it;

aachavarria committed:
adding missing unsubscribe

aachavarria committed:
fixing ids

aachavarria committed:
renaming updateDetailedItem to reloadDetailedItem;

aachavarria committed:
PR requested changes;

aachavarria committed:
changing items.byId for byPath

aachavarria committed:
PR requested Changes

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/permissions

aachavarria committed:
adding pathNavigator items to state;

aachavarria committed:
removing items from pathNavigator.ts state;

rart commented on ui/app/src/services/content.ts:
```diff
@@ -1463,16 +1464,16 @@ export function getChildrenByPath(
-export function copy(site: string, item: Partial<LegacyItem>): Observable<{ success: boolean }> {
-  let _item = item.children ? { item: [item] } : { item: [{ uri: item.path }] };
+export function copy(site: string, path: string, copyItem?: CopyItem): Observable<{ success: boolean }> {
+  let item = path ? { item: [{ uri: path }] } : { item: [copyItem] };
```
Looks like you’d be sending [undefined] when no copyitem is supplied

aachavarria replied:
yes, CopyItem is not needed when you already sent a path

rart commented on ui/app/src/services/content.ts:
```diff
@@ -1482,27 +1483,33 @@ export function cut(site: string, item: SandboxItem): Observable<any> {
+  }
+  parentPath = parentPath.split('/');
+  parentPath.pop();
+  parentPath = parentPath.join('/');
```
Create a getParentPath útil and use here

aachavarria replied:
done

rart commented on ui/app/src/services/content.ts:
```diff
@@ -1482,27 +1483,33 @@ export function cut(site: string, item: SandboxItem): Observable<any> {
 }
 
-export function getPages(site: string, item: any): Observable<any> {
+export function getPages(site: string, path: string): Observable<LegacyItem> {
```
What’s this for?

aachavarria replied:
this is for the copyDialog, this uses this service to get the selected page and the subpages/children, etc

rart commented on ui/app/src/utils/content.ts:
```diff
@@ -165,6 +176,28 @@ export function parseLegacyItemToDetailedItem(item: LegacyItem | LegacyItem[]):
+export function parseSandBoxItemToDetailedItem(item: SandboxItem[]): DetailedItem[];
+export function parseSandBoxItemToDetailedItem(item: SandboxItem | SandboxItem[]): DetailedItem | DetailedItem[] {
+  if (Array.isArray(item)) {
+    // including level descriptors to avoid issues on pathNavigator;
```
What’s the issue with level descriptors?

aachavarria replied:
the pathNavigator is including levelDescriptor on the list of pages, and the levelDescriptor was not included on the Store

rart commented on ui/app/src/utils/content.ts:
```diff
@@ -165,6 +176,28 @@ export function parseLegacyItemToDetailedItem(item: LegacyItem | LegacyItem[]):
+export function parseSandBoxItemToDetailedItem(item: SandboxItem | SandboxItem[]): DetailedItem | DetailedItem[] {
+  if (Array.isArray(item)) {
+    // including level descriptors to avoid issues on pathNavigator;
+    return item.flatMap(i => [parseSandBoxItemToDetailedItem(i)]);
```
there’s no filtering happening here so you can use map instead of flatMap

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -0,0 +1,489 @@
+/*
```
file has pending todos

aachavarria replied:
not sure what are we going to do with the embedded items, using this menu.

aachavarria committed:
updating partial detailed item on ItemMenu.tsx;

aachavarria committed:
Fixes on Copy service to receive path or copyItem;
Revert on ItemMenu.tsx;
using completeDetailedItem on Widget.tsx;

aachavarria committed:
renaming getPages to getLegacyItemsTree;

aachavarria committed:
fixes on reading localEditMode;

aachavarria committed:
ItemMenu.tsx: exporting Loader;
Widget.tsx: Fix on loaderItems;
MediaCard.tsx: adding Asset Preview;
Search: adding new SimpleMenu to search; Changes on Edit to use LegacyForms;

aachavarria committed:
MediaCard.tsx: Modernize callback Functions;
Search.tsx: addigng onPreview and onEdit;

aachavarria committed:
removing unnecessary props from search.js

aachavarria committed:
fixing ace-mode

aachavarria committed:
search.js: deleting old code;
adding getPreviewURLFromPath to Search.tsx;
changes on getPreviewURLFromPath to remove index.html and return only the path;

rart committed:
Update PreviewConcierge.tsx

aachavarria committed:
adding epic to listen pathNavigatorSetKeyword;
renaming copyDialog;

aachavarria committed:
Merge remote-tracking branch 'aaron/feature/permissions' into feature/permissions

aachavarria committed:
adding missing rename itemMenu Dispatch;
renameValidation;

rart reviewed (changes requested):
Fix codacy review issues

aachavarria committed:
fix code issues;

aachavarria committed:
addinng missing type

sumerjabri commented:
@alhambrav please fix conflicts

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/permissions

 Conflicts:
	ui/app/src/components/SystemStatus/GlobalDialogManager.tsx
	ui/app/src/models/GlobalState.ts
	ui/app/src/state/actions/dialogs.ts
	ui/app/src/state/reducers/dialogs/index.ts

aachavarria commented:
@sumerjabri done.
